File: androidApp/src/main/kotlin/com/example/ta33/ui/components/KPRow.kt
package com.example.ta33.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import com.example.ta33.domain.model.ControlPointState
import com.example.ta33.ui.theme.Ta33Theme

/**
 * Jeden řádek deníku: 56dp stavový swatch + titul (bodyStrong) + podtitul.
 * Barvy swatche a chování řádku odvozeny z [state] (a [isFinish]) přes `Ta33Theme.colors`.
 */
@Composable
fun KPRow(
    ordinal: Int,
    title: String,
    subtitle: String,
    state: ControlPointState,
    modifier: Modifier = Modifier,
    isFinish: Boolean = false,
) {
    val isLocked = state == ControlPointState.LOCKED
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .alpha(if (isLocked) 0.55f else 1f)
            .shadow(Ta33Theme.spacing.x1, Ta33Theme.radius.lg),
        color = MaterialTheme.colorScheme.surface,
        shape = Ta33Theme.radius.lg,
    ) {
        Row(
            modifier = Modifier.padding(Ta33Theme.spacing.x3),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Ta33Theme.spacing.x3),
        ) {
            Swatch(ordinal = ordinal, state = state, isFinish = isFinish)
            Column(
                modifier = Modifier.padding(vertical = Ta33Theme.spacing.x1),
                verticalArrangement = Arrangement.spacedBy(Ta33Theme.spacing.x1),
            ) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleMedium,
                    color = Ta33Theme.colors.fgStrong,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    text = subtitle,
                    style = MaterialTheme.typography.bodyMedium,
                    color = Ta33Theme.colors.fgMuted,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

@Composable
private fun Swatch(
    ordinal: Int,
    state: ControlPointState,
    isFinish: Boolean,
) {
    val colors = Ta33Theme.colors
    val bg: Color
    val fg: Color
    val glow: Boolean
    when (state) {
        ControlPointState.DONE -> {
            bg = colors.kpDoneBg; fg = colors.kpDoneFg; glow = false
        }
        ControlPointState.ACTIVE -> {
            bg = colors.kpActiveBg; fg = colors.kpActiveFg; glow = true
        }
        ControlPointState.FINISH -> {
            bg = colors.kpFinishBg; fg = colors.kpFinishFg; glow = false
        }
        ControlPointState.LOCKED -> {
            bg = colors.kpLockedBg; fg = colors.kpLockedFg; glow = false
        }
    }
    val shape = Ta33Theme.radius.md
    Surface(
        modifier = Modifier
            .size(Ta33Theme.spacing.x9)
            .then(
                if (glow) {
                    Modifier.shadow(
                        elevation = Ta33Theme.spacing.x2,
                        shape = shape,
                        ambientColor = colors.kpActiveBg,
                        spotColor = colors.kpActiveBg,
                    )
                } else {
                    Modifier
                },
            ),
        color = bg,
        shape = shape,
    ) {
        Box(contentAlignment = Alignment.Center) {
            when {
                state == ControlPointState.DONE -> Icon(
                    imageVector = Ta33Icons.Check,
                    contentDescription = null,
                    tint = fg,
                    modifier = Modifier.size(Ta33Theme.spacing.x6),
                )
                isFinish || state == ControlPointState.FINISH -> Icon(
                    imageVector = Ta33Icons.Star,
                    contentDescription = null,
                    tint = fg,
                    modifier = Modifier.size(Ta33Theme.spacing.x6),
                )
                else -> Text(
                    text = ordinal.toString(),
                    style = MaterialTheme.typography.displaySmall,
                    color = fg,
                )
            }
        }
    }
}

@Preview
@Composable
private fun KPRowStatesPreview() {
    Ta33Theme {
        Column(
            modifier = Modifier.padding(Ta33Theme.spacing.x5),
            verticalArrangement = Arrangement.spacedBy(Ta33Theme.spacing.x3),
        ) {
            KPRow(
                ordinal = 1,
                title = "KP-01 · Start",
                subtitle = "Splněno · 08:14",
                state = ControlPointState.DONE,
            )
            KPRow(
                ordinal = 2,
                title = "KP-02 · Sloní pramen",
                subtitle = "Další",
                state = ControlPointState.ACTIVE,
            )
            KPRow(
                ordinal = 3,
                title = "KP-03 · Vyhlídka",
                subtitle = "Zamčeno",
                state = ControlPointState.LOCKED,
            )
            KPRow(
                ordinal = 5,
                title = "Cíl · Adršpach",
                subtitle = "Zamčeno",
                state = ControlPointState.LOCKED,
                isFinish = true,
            )
            KPRow(
                ordinal = 5,
                title = "Cíl · Adršpach",
                subtitle = "Splněno · 11:02",
                state = ControlPointState.FINISH,
                isFinish = true,
            )
        }
    }
}

File: iosApp/iosApp/UI/Components/KPRow.swift
import SwiftUI
import Shared

/// Jeden řádek deníku: 56pt stavový swatch + titul (bodyStrong) + podtitul (small).
/// Barvy swatche a chování řádku odvozeny z [state] (a [isFinish]) přes `Ta33Color`.
/// Zrcadlí Android `KPRow`; ikony jsou SF Symbols (`checkmark`, `star.fill`).
struct KPRow: View {
    let ordinal: Int32
    let title: String
    let subtitle: String
    let state: ControlPointState
    var isFinish: Bool = false

    var body: some View {
        HStack(spacing: Ta33Spacing.x3) {
            swatch
            VStack(alignment: .leading, spacing: Ta33Spacing.x1) {
                Text(title)
                    .font(Ta33Font.bodyStrong)
                    .foregroundStyle(Ta33Color.fgStrong)
                    .lineLimit(1)
                Text(subtitle)
                    .font(Ta33Font.small)
                    .foregroundStyle(Ta33Color.fgMuted)
                    .lineLimit(1)
            }
            Spacer(minLength: 0)
        }
        .padding(Ta33Spacing.x3)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Ta33Color.paper)
        .clipShape(RoundedRectangle(cornerRadius: Ta33Radius.lg, style: .continuous))
        .shadow(color: .black.opacity(0.06), radius: Ta33Spacing.x1, x: 0, y: Ta33Spacing.x1)
        .opacity(state == .locked ? 0.55 : 1.0)
    }

    @ViewBuilder private var swatch: some View {
        let style = swatchStyle
        ZStack {
            RoundedRectangle(cornerRadius: Ta33Radius.md, style: .continuous)
                .fill(style.bg)
            swatchContent(fg: style.fg)
        }
        .frame(width: Ta33Spacing.x9, height: Ta33Spacing.x9)
        .shadow(
            color: style.glow ? Ta33Color.orange.opacity(0.45) : .clear,
            radius: style.glow ? Ta33Spacing.x2 : 0,
            x: 0,
            y: style.glow ? Ta33Spacing.x1 : 0
        )
    }

    private var swatchStyle: (bg: Color, fg: Color, glow: Bool) {
        switch state {
        case .done: return (Ta33Color.kpDoneBg, Ta33Color.kpDoneFg, false)
        case .active: return (Ta33Color.kpActiveBg, Ta33Color.kpActiveFg, true)
        case .finish: return (Ta33Color.kpFinishBg, Ta33Color.kpFinishFg, false)
        case .locked: return (Ta33Color.kpLockedBg, Ta33Color.kpLockedFg, false)
        }
    }

    @ViewBuilder private func swatchContent(fg: Color) -> some View {
        if state == .done {
            Image(systemName: "checkmark")
                .font(.system(size: Ta33Spacing.x6, weight: .bold))
                .foregroundStyle(fg)
        } else if isFinish || state == .finish {
            Image(systemName: "star.fill")
                .font(.system(size: Ta33Spacing.x5))
                .foregroundStyle(fg)
        } else {
            Text("\(ordinal)")
                .font(Ta33Font.display3)
                .foregroundStyle(fg)
        }
    }
}

#Preview("KPRow states") {
    VStack(spacing: Ta33Spacing.x3) {
        KPRow(ordinal: 1, title: "KP-01 · Start", subtitle: "Splněno · 08:14", state: .done)
        KPRow(ordinal: 2, title: "KP-02 · Sloní pramen", subtitle: "Další", state: .active)
        KPRow(ordinal: 3, title: "KP-03 · Vyhlídka", subtitle: "Zamčeno", state: .locked)
        KPRow(ordinal: 5, title: "Cíl · Adršpach", subtitle: "Zamčeno", state: .locked, isFinish: true)
        KPRow(ordinal: 5, title: "Cíl · Adršpach", subtitle: "Splněno · 11:02", state: .finish, isFinish: true)
    }
    .padding(Ta33Spacing.x5)
    .background(Ta33Color.cream)
}
